stablehlo/reference/Element.cpp

```diff
@@ -720,6 +720,57 @@ Element rsqrt(const Element &el) {
       [](std::complex<double> e) { return 1.0 / std::sqrt(e); });
 }
 
+Element sign(const Element &el) {
+  Type type = el.getType();
+
+  if (isSupportedIntegerType(type)) {
+    auto intEl = el.getIntegerValue();
+    if (intEl.isZero()) return Element(type, (int64_t)0);
+    if (intEl.isNegative())
+      return Element(type, (int64_t)-1);
+    else
+      return Element(type, (int64_t)1);
+  }
+
+  if (isSupportedFloatType(type)) {
+    auto elVal = el.getFloatValue();
+
+    if (elVal.isNaN()) return Element(el);
+    if (elVal.isNegZero()) return Element(type, (double)-0.0);
+    if (elVal.isPosZero()) return Element(type, (double)0.0);
+    if (elVal.isNegative()) return Element(type, (double)-1.0);
+    if (!elVal.isNegative()) return Element(type, (double)1.0);
+  }
+
+  if (isSupportedComplexType(type)) {
+    auto elVal = el.getComplexValue();
+    const llvm::fltSemantics &elSemantics = elVal.real().getSemantics();
+
+    if (elVal.real().isNaN() || elVal.imag().isNaN())
+      return Element(type,
+                     std::complex<APFloat>(APFloat::getNaN(elSemantics),
+                                           APFloat::getZero(elSemantics)));
+
+    if (elVal.real().isZero() && elVal.imag().isZero())
+      return Element(type,
+                     std::complex<APFloat>(APFloat::getZero(elSemantics),
+                                           APFloat::getZero(elSemantics)));
```

Comment:
Also, how about we drop the check for whether el is equal to zero to simplify the implementation?

The spec suggests that this is implementation-defined behavior, so we are free to do whatever we want here: "StableHLO programs are validated through an extensive set of constraints for individual ops, which rules out many classes of errors prior to run time. However, error conditions are still possible, e.g. through integer overflows, out-of-bounds accesses, etc. Unless explicitly called out, all these errors result in implementation-defined behavior, but this may change in the future".

Reply:
Huh interesting it looks like this actually might need to be part of the spec: https://github.com/openxla/stablehlo/pull/1395#discussion_r1164397409. Let's discuss there, update the spec if needed and then reproduce this in the interpreter.

Reply:
Hello @burmako ,

`stablehlo/testdata/sign_special_0_dtype_complex64.mlir` defines `stablehlo.sign(0+0i)` outputs `0+0i`. But the suggested implementation returns `NaN + NaN i`.

This is the only test failing in CI. Any suggestion on what I should do ?

Reply:
How about we disable the test for now and take care of it in #1395?

Reply:
Sure, that works.